components/codec/src/bits/buffered_read.rs

```diff
@@ -0,0 +1,201 @@
+// Copyright 2023 The CeresDB Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::bits::{Bit, BIT_MASKS};
+
+/// BufferedReader
+/// BufferedReader encapsulates a buffer of bytes which can be read from.
+#[derive(Debug)]
+pub struct BufferedReader<'a> {
+    /// internal buffer of bytes
+    bytes: &'a [u8],
+    /// index into bytes
+    byte_idx: usize,
+    /// position in the byte we are currently reading
+    bit_idx: u32,
+}
+
+impl<'a> BufferedReader<'a> {
+    /// new creates a new `BufferedReader` from `bytes`
+    pub fn new(bytes: &'a [u8]) -> Self {
+        BufferedReader {
+            bytes,
+            byte_idx: 0,
+            bit_idx: 0,
+        }
+    }
+
+    fn get_byte(&self) -> Option<u8> {
+        if self.is_eof() {
+            return None;
+        }
+        self.bytes.get(self.byte_idx).cloned()
+    }
+
+    #[inline]
+    fn is_eof(&self) -> bool {
+        self.byte_idx >= self.bytes.len()
+    }
+
+    fn advance_one_bit(&mut self) {
+        if self.bit_idx == 7 {
+            self.bit_idx = 0;
+            self.byte_idx += 1;
+        } else {
+            self.bit_idx += 1;
+        }
+    }
+}
+
+impl<'a> BufferedReader<'a> {
+    pub fn next_bit(&mut self) -> Option<Bit> {
+        if self.is_eof() {
+            return None;
+        }
+        let byte = self.get_byte().unwrap();
+        let bit = Bit(u8::from(byte & BIT_MASKS[self.bit_idx as usize] != 0));
+        self.advance_one_bit();
+        Some(bit)
+    }
+
+    fn next_byte(&mut self) -> Option<u8> {
+        if self.bit_idx == 0 {
+            let byte = self.get_byte();
+            self.byte_idx += 1;
+            return byte;
+        }
+
+        let mut byte = 0;
+        let mut b = match self.get_byte() {
+            None => return None,
+            Some(b) => b,
+        };
+
+        byte |= b.wrapping_shl(self.bit_idx);
+
+        self.byte_idx += 1;
+        b = match self.get_byte() {
+            None => return None,
+            Some(b) => b,
+        };
+
+        byte |= b.wrapping_shr(8 - self.bit_idx);
+
+        Some(byte)
+    }
+
+    /// Fetch the next `num` bits, and advance the inner cursor. And the
+    /// returned bits are stored in a u64 integer in little-endian order.
+    ///
+    /// Example: the returned value will be `0x0000 0000 0000 000F` if 4 bits is
+    /// fetched and they are all set.
+    pub fn next_bits(&mut self, mut num: u32) -> Option<u64> {
```

Comment:
It may perform better if follow these steps:
1. Read some bits to make the `bit_idx` == 0;
2. Read the remaining bits byte by byte;
3. Read the remaining bits which can't form a byte;

If this implementation is adopted, `next_byte` can be implemented more simply.